File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/Python Programs/3_complete_linked_list.py
class Node:
    def __init__(self, d): self.data, self.next = d, None

class LL:
    def __init__(self): self.head = None
    def iaf(self, d): n = Node(d); n.next = self.head; self.head = n
    def ial(self, d): n = Node(d); (setattr(self, 'head', n) if not self.head else [t for t in [self.head] if not [setattr(t, 'next', n) for t in iter(lambda x=self.head: x.next if x.next else None, None)]])
    def iaa(self, d, p): self.iaf(d) if p == 1 else (lambda t: [setattr(t, 'next', Node(d)), setattr(Node(d), 'next', t.next)] if t else self.ial(d))([self.head] + [self.head := self.head.next for _ in range(p-2) if self.head and self.head.next][-1:] or [None])[0] if self.head else self.ial(d)
    def df(self): self.head = self.head.next if self.head else None
    def dl(self): (setattr(self, 'head', None) if not self.head.next else [setattr(t, 'next', None) for t in iter(lambda x=self.head: x.next if x.next.next else None, None)]) if self.head else None
    def da(self, p): self.df() if p == 1 else ([setattr(t, 'next', t.next.next) for t in [self.head] + [self.head := self.head.next for _ in range(p-2) if self.head and self.head.next] if t and t.next] or None)
    def disp(self): print(' -> '.join([str(t.data) for t in iter(lambda x=[self.head]: x[0] and (x.append(x[0].next), x.pop(0))[1] or None, None)]) + ' -> None' if self.head else 'Empty')

ll = LL()
while True:
    c = input("1:IF 2:IL 3:IA 4:DF 5:DL 6:DA 7:Show 8:Exit: ")
    if c == '1': ll.iaf(int(input()))
    elif c == '2': ll.ial(int(input()))
    elif c == '3': ll.iaa(int(input()), int(input()))
    elif c == '4': ll.df()
    elif c == '5': ll.dl()
    elif c == '6': ll.da(int(input()))
    elif c == '7': ll.disp()
    elif c == '8': break

File: Week-Wise-Assignments/Assignment 3/Compressed Solutions/C Programs/3_complete_linked_list.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node { int data; struct Node* next; } Node;
Node* cn(int d) { Node* n = malloc(sizeof(Node)); n->data = d; n->next = NULL; return n; }
Node* iaf(Node* h, int d) { Node* n = cn(d); n->next = h; return n; }
Node* ial(Node* h, int d) { Node* n = cn(d); if (!h) return n; Node* t = h; while (t->next) t = t->next; t->next = n; return h; }
Node* iaa(Node* h, int d, int p) { if (p == 1) return iaf(h, d); Node* n = cn(d), *t = h; for (int i = 1; i < p-1 && t; i++) t = t->next; if (!t) return ial(h, d); n->next = t->next; t->next = n; return h; }
Node* df(Node* h) { if (!h) return NULL; Node* t = h; h = h->next; free(t); return h; }
Node* dl(Node* h) { if (!h || !h->next) { free(h); return NULL; } Node* t = h; while (t->next->next) t = t->next; free(t->next); t->next = NULL; return h; }
Node* da(Node* h, int p) { if (p == 1) return df(h); Node* t = h; for (int i = 1; i < p-1 && t; i++) t = t->next; if (!t || !t->next) return h; Node* d = t->next; t->next = d->next; free(d); return h; }
void disp(Node* h) { while (h) printf("%d -> ", h->data), h = h->next; printf("NULL\n"); }
int main() {
    Node* h = NULL; int c, d, p;
    while (scanf("%d", &c) && c != 11) {
        if (c == 1) scanf("%d", &d), h = iaf(h, d);
        else if (c == 2) scanf("%d", &d), h = ial(h, d);
        else if (c == 3) scanf("%d%d", &d, &p), h = iaa(h, d, p);
        else if (c == 4) h = df(h);
        else if (c == 5) h = dl(h);
        else if (c == 6) scanf("%d", &p), h = da(h, p);
        else if (c == 7) disp(h);
    }
    return 0;
}